doc/llvm_obfuscation_tutorial.rst

```diff
@@ -0,0 +1,968 @@
+============================================================
+Introduction on turning good code into attrocities with LLVM
+============================================================
+
+Objective
+=========
+The objective of this post is to explain the basics of LLVM bytecode obfuscation.
+*'But why obfuscate the LLVM bytecode? Why not the source code? Or the binary?'*, you may ask.
+
+Well it's because LLVM is *super swag* right now and using it is cool.
+But, regarding the engineering aspects, it is because there are lots of front-ends converting different languages into the same LLVM bytecode (Clang/Clang++ for C/C++, Mono LLVM for C#, Pyston for ... Python and so on).
+Hence by working at the bytecode level we can obfuscate programs produced by many languages without even knowing them.
+Another good thing is that the obfuscation can be easily integrated with the compilation chain.
+
+Now let's talk about what we're going to do.
+Our mission (and we have no choice but to accept it) is to obfuscate all null literals in the code.
+It means that we are going to replace (almost) all the zeroes in the code by a non-trivial boolean expression always false.
+
+        prime1 * ((x | any1)**2) != prime2 * ((y | any2)**2)
+
+    - prime1 and prime2 are *distinct* prime numbers
+    - any1 and any2 are *distinct* strictly positive random numbers
+    - x and y are two variables picked from the program (they have to be reachable from the obfuscation instructions)
+
+This expression will always return a boolean zero (false). The idea is to insert this test into our code, just before the 0 we want to obfuscate and to replace this 0 by the result of our comparison.
+As you have probably noticed we will have to pay attention to the type of the original 0 and make sure we cast the result of our expression to its type.
+
+This obfuscation may not be the most sophisticated ever written but it's enough to learn the basics of LLVM bytecode obfuscation (and to annoy our reverse engineer friends.)
```

Comment:
and maybe to annoy our .. firends for a few minutes, until they issue a miasm script to clean this a bit :-)